library/gui/key_bindings/src/main/java/org/quiltmc/qsl/key/bindings/impl/KeyBindingRegistryImpl.java

```
@@ -0,0 +1,134 @@
+import org.apache.logging.log4j.Logger;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+
+import net.minecraft.client.option.GameOptions;
+import net.minecraft.client.option.KeyBinding;
+
+import org.quiltmc.qsl.key.bindings.mixin.client.KeyBindingAccessor;
+
+@Environment(EnvType.CLIENT)
+public class KeyBindingRegistryImpl {
+	public static final Logger LOGGER = LogManager.getLogger();
+
+	private static Map<KeyBinding, Boolean> quiltKeys = new HashMap<>();
+	private static List<KeyBindingManager> keyBindingManagers = new ArrayList<>(1);
+	private static KeyBinding[] enabledQuiltKeysArray = new KeyBinding[] {};
+	private static List<KeyBinding> disabledQuiltKeys = new ArrayList<>(0);
+
+	public static KeyBinding registerKeyBinding(KeyBinding key, boolean enabled) {
+		if (key == null) {
+			LOGGER.error("Attempted to register a null key bind!");
+			return null;
+		}
+
+		for (KeyBinding otherKey : quiltKeys.keySet()) {
+			if (key == otherKey) {
+				LOGGER.error("{} has already been registered!", key.getTranslationKey());
+				return null;
+			} else if (key.getTranslationKey() == otherKey.getTranslationKey()) {
+				LOGGER.error("Attempted to register {}, but a key bind with the same translation key has already been registered!", key.getTranslationKey());
+				return null;
+			}
+		}
+
+		quiltKeys.put(key, enabled);
+		applyChanges();
+		if (!enabled) {
+			KeyBindingAccessor.getKeysById().remove(key.getTranslationKey());
+		}
+
+		return key;
+	}
+
+	public static KeyBinding getKeyBinding(String translationKey) {
+		for (KeyBinding key : quiltKeys.keySet()) {
+			if (key.getTranslationKey() == translationKey) {
+				return key;
+			}
+		}
+
+		return null;
+	}
+
+	public static boolean isEnabled(KeyBinding key) {
+		if (quiltKeys.containsKey(key)) {
+			return quiltKeys.get(key);
+		}
+
+		return false;
+	}
+
+	public static void setEnabled(KeyBinding key, boolean enabled) {
+		if (quiltKeys.containsKey(key)) {
```

This should throw an exception if the key isn't a quilt key

---

Good idea

---

And done; It's currently the same exception for both non-registered keys and vanilla ones, but I'll eventually split it